pkg/alertmanager/multitenant.go

```diff
@@ -737,6 +766,59 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
 		return
 	}
 
+	if am.cfg.ShardingEnabled && am.distributor.IsPathSupported(req.URL.Path) {
+		am.distributor.DistributeRequest(w, req)
+		return
+	}
+
+	// If sharding is not enabled or Distributor does not support this path,
+	// it is a simple pass-through.
+	am.serveHTTP(w, req)
+}
+
+// HandleRequest serves the Alertmanager's web UI and API sent via gRPC.
+func (am *MultitenantAlertmanager) HandleRequest(ctx context.Context, in *alertmanagerpb.Request) (*alertmanagerpb.Response, error) {
+	req, err := http.NewRequestWithContext(ctx, in.HttpRequest.Method, in.HttpRequest.Url, bytes.NewReader(in.HttpRequest.Body))
+	if err != nil {
+		return nil, err
+	}
+
+	w := httptest.NewRecorder()
+	am.serveHTTP(w, req)
+
+	// Convert http response to gRPC response.
+
+	var body []byte
+	httpResp := w.Result()
+	if httpResp.Body != nil {
+		body, err = ioutil.ReadAll(httpResp.Body)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// All non 200 status are sent as errors.
```

Comment:
Why? Shouldn't we just forward the HTTP response "as is", whatever is the response code? Also you're checking for `== 200` but potentially there may be other successful codes (eg. 2xx, 3xx).